File: android/app/src/main/java/dev/termland/android/data/HostProfile.kt
package dev.termland.android.data

import dev.termland.core.MobileCodec
import dev.termland.core.ServerProfile
import dev.termland.core.SessionParams
import kotlinx.serialization.Serializable
import java.util.UUID

/**
 * A saved server, plus the session defaults the user last chose for it.
 *
 * This is the app's own persistence model, deliberately separate from the core's
 * [ServerProfile] record: it carries UI-only fields (label, id, rememberPassword)
 * and survives contract changes on the Rust side.
 */
@Serializable
data class HostProfile(
    val id: String = UUID.randomUUID().toString(),
    val label: String = "",
    val host: String = "",
    val port: Int = DEFAULT_PORT,
    val useTls: Boolean = true,
    val acceptInvalidCerts: Boolean = false,
    val username: String = "",
    /**
     * Stored in clear in DataStore only when [rememberPassword] is set.
     *
     * TODO(M4): move this to an Android Keystore-wrapped blob, as
     * docs/mobile-clients.md calls for. Until then the honest default is
     * "don't persist it" and prompt at connect time.
     */
    val password: String = "",
    val rememberPassword: Boolean = false,
    // --- session defaults ---
    val quality: Int = 75,
    val audio: Boolean = false,
    val desktopShell: String = "",
    val appCommand: String = "",
) {
    val displayName: String get() = label.ifBlank { host.ifBlank { "New server" } }

    val subtitle: String
        get() = buildString {
            if (username.isNotBlank()) append(username).append('@')
            append(host.ifBlank { "?" }).append(':').append(port)
            append(if (useTls) "  TLS" else "  plain")
            if (useTls && acceptInvalidCerts) append(" (unverified)")
        }

    companion object {
        const val DEFAULT_PORT = 7867
    }
}

/**
 * @param passwordOverride a password typed at connect time when the profile does
 *        not persist one.
 */
fun HostProfile.toServerProfile(passwordOverride: String? = null): ServerProfile = ServerProfile(
    host = host.trim(),
    port = port.toUShort(),
    useTls = useTls,
    acceptInvalidCerts = acceptInvalidCerts,
    username = username.trim().ifBlank { null },
    password = (passwordOverride ?: password).ifBlank { null },
)

/**
 * Build [SessionParams] for this profile at a concrete surface size.
 *
 * [supportedCodecs] must be the *probed* decoder set — see
 * [dev.termland.android.net.CodecSupport]. Advertising codecs the device cannot
 * decode is the one way to break the server-side negotiation.
 */
fun HostProfile.toSessionParams(
    width: Int,
    height: Int,
    supportedCodecs: List<MobileCodec>,
): SessionParams = SessionParams(
    width = width.toUInt(),
    height = height.toUInt(),
    quality = quality.coerceIn(1, 100).toUByte(),
    audio = audio,
    desktopShell = desktopShell.trim().ifBlank { null },
    appCommand = appCommand.trim().ifBlank { null },
    supportedCodecs = supportedCodecs,
)

File: ios/Termland/HostProfile.swift
import Foundation

/// UI-owned profile data. Credentials deliberately do not live in this
/// Codable record: only KeychainStore stores the password, indexed by `id`.
struct HostProfile: Identifiable, Codable, Equatable {
    var id = UUID()
    var label = ""
    var host = ""
    var port = 7867
    var useTLS = true
    var acceptInvalidCertificates = false
    var username = ""

    var displayName: String {
        label.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? host : label
    }

    var subtitle: String {
        let identity = username.trimmingCharacters(in: .whitespacesAndNewlines)
        let endpoint = "\(host):\(port)"
        let transport = useTLS ? "TLS" : "Plain TCP"
        let verification = useTLS && acceptInvalidCertificates ? " · unverified" : ""
        return identity.isEmpty ? "\(endpoint) · \(transport)\(verification)" : "\(identity)@\(endpoint) · \(transport)\(verification)"
    }

    func coreProfile(password: String?) -> ServerProfile {
        ServerProfile(
            host: host.trimmingCharacters(in: .whitespacesAndNewlines),
            port: UInt16(clamping: port),
            useTls: useTLS,
            acceptInvalidCerts: acceptInvalidCertificates,
            username: username.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
            password: password?.nilIfEmpty,
            useSsh: false,
            useQuic: false
        )
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}
